costmap_2d/plugins/voronoi_inflation_layer.cpp

```diff
@@ -0,0 +1,459 @@
+      unsigned int sy = current_cell.src_y_;
+
+      if (std::abs((int)mx - (int)sx) > cell_inflation_radius_ + 2) {
+        ROS_INFO("od: %f, vd: %f, mx: %d, my %d, sx: %d, sy %d",
+         obs_dist[index], vor_dist[index],
+         mx, my, sx, sy);
+      }
+
+      // Calculate the cost for the final cost value.
+      unsigned char cost = costLookup(obs_dist[index], vor_dist[index]);
+      unsigned char old_cost = master_array[index];
+      if (old_cost == NO_INFORMATION && cost >= INSCRIBED_INFLATED_OBSTACLE)
+        master_array[index] = cost;
+      else
+        master_array[index] = std::max(old_cost, cost);
+
+      if (dist_bin.first > cell_inflation_radius_ + 1)
+      {
+        continue;
+      }
+
+
+
+      // attempt to put the neighbors of the current cell onto the inflation list
+      if (mx > 0)
+        enqueue(index - 1, mx - 1, my, sx, sy, vor_cells);
+      if (my > 0)
+        enqueue(index - size_x, mx, my - 1, sx, sy, vor_cells);
+      if (mx < size_x - 1)
+        enqueue(index + 1, mx + 1, my, sx, sy, vor_cells);
+      if (my < size_y - 1)
+        enqueue(index + size_x, mx, my + 1, sx, sy, vor_cells);
+    }
+  }
+  vor_cells.clear();
+
+  // Clean in all up.
+  delete[] obs_dist;
+  delete[] vor_dist;
+
+ROS_INFO("Finished voronoi cells.");
+}
+
+/**
+ * @brief  Given an index of a cell in the costmap, place it into a list pending for obstacle inflation
+ * @param  grid The costmap
+ * @param  index The index of the cell
+ * @param  mx The x coordinate of the cell (can be computed from the index, but saves time to store it)
+ * @param  my The y coordinate of the cell (can be computed from the index, but saves time to store it)
+ * @param  src_x The x index of the obstacle point inflation started at
+ * @param  src_y The y index of the obstacle point inflation started at
+ */
+inline void VoronoiInflationLayer::enqueue(unsigned int index, unsigned int mx, unsigned int my,
+                                    unsigned int src_x, unsigned int src_y,
+                                    std::map<double, std::vector<CellData> >& queue_map)
+{
+  // we compute our distance table one cell further than the inflation radius dictates so we can make the check below
+  double distance = distanceLookup(mx, my, src_x, src_y);
+
+
+  // push the cell data onto the inflation list and mark
+  queue_map[distance].push_back(CellData(index, mx, my, src_x, src_y));
+}
+
+void VoronoiInflationLayer::computeCaches()
+{
+
+  ROS_INFO_STREAM("Cache compute with ir: " << inflation_radius_ << ", cir: "
+    << cell_inflation_radius_ << ", w: " << weight_
+    << ", insc: " << inscribed_radius_ << ", res: " << resolution_);
+  if (cell_inflation_radius_ == 0)
+    return;
+
+  // based on the inflation radius... compute distance and cost caches
+  if (cell_inflation_radius_ != cached_cell_inflation_radius_)
+  {
+    deleteKernels();
+
+    cached_costs_.clear();
+
+    cached_distances_ = new double*[cell_inflation_radius_ + 5];
```

Comment:
If this is something from other layers, it may not make sense.  You can create a shared_ptr of shared_ptr but maybe it's not worth the time.